### PR TITLE
Harden Client API and Swarm abort cleanup

### LINKED ISSUE #5099: Late Client API results can orphan external trainer processes during CJ teardown
## Problem

For external Client API execution with `launch_once=false`, a trainer that publishes a late result can outlive its owning Client Job process when the server has already accepted enough client responses.

The result source remains live while the backend result reaper waits. If that wait exceeds the outer CJ teardown budget, the CJ is terminated before it reaches owned-process cleanup and the trainer process group becomes orphaned.

## Reproduction

1. Run three clients with a server response threshold of two.
2. Use per-task external trainers with `launch_once=false`.
3. Delay one client result until after the server accepts the other two responses.
4. End the run and clean the job workspace.
5. Audit processes after framework teardown.

## Observed behavior

Affected trainers survived with PPID 1, retained a deleted job working directory, and held a CLOSE-WAIT Cell socket. No process-group termination marker was emitted for those launches before the CJ exited.

The existing result-source wait could be longer than the actual outer CJ teardown budget. Normal finalization therefore did not guarantee that all trainer groups owned by the CJ were dead before the CJ disappeared.

## Expected behavior

- END_RUN completes within the CJ teardown budget.
- Every trainer process group owned by the CJ is terminated before finalization returns, including retired launches with late or unconsumed result sources.
- Abrupt CJ death cannot leave an external trainer orphaned.
- Attach trainers are unaffected because their process lifecycle is externally owned.

## Acceptance criteria

- Bound live accepted-result cleanup to a short END_RUN acknowledgement interval.
- After that interval, force-clean all still-owned active and retired trainer groups without starting another long termination grace.
- Detect abrupt owner-CJ disappearance from an external trainer and terminate its isolated process group.
- Do not apply owner-PID monitoring to Attach mode.
- Add deterministic late-result and owner-death regression tests.
- Run immediate and delayed leak audits after CPU and GPU Client API integration jobs.

## Proposed fix

Draft PR #5097 implements bounded result-reaper cleanup, exact owned-process-group termination, and the external trainer owner watchdog.

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
## Summary

Fix #5099.

- Bound live accepted-result cleanup to the END_RUN acknowledgement interval instead of the longer streaming timeout.
- Force-stop every still-owned trainer process group after that bound without adding a second termination grace.
- Pass the owning Client Job PID to launched trainers and monitor it so abrupt CJ exit cannot orphan the trainer process group.
- Propagate graceful job abort to an external backend even after `execute()` returned an accepted lazy result, so aborted teardown does not preserve a live result source.
- Fail a job promptly when an owned external trainer dies after its lazy result was accepted, interrupting the exact receiver download instead of entering the 600-second retry path.
- Preserve user-abort precedence so late process exits cannot reclassify `FINISHED:ABORTED` as an execution failure.
- Order result archival, terminal failure reporting, F3 shutdown, and process cleanup so terminal state is authenticated and authoritative.
- Make active CCWF work abort at `ABOUT_TO_END_RUN`, including work not represented in `ClientRunner.running_tasks`.
- Carry the run abort signal into Swarm's explicit lazy-reference decode, cancel active disk tensor consumers under the job-scoped root, serialize cleanup with chunk writes, and reject writes after cleanup.
- Add a capability-negotiated receiver cancellation handshake so an abandoned download settles the producer's accepted result transaction promptly, including all sibling refs for that receiver.
- Keep Attach mode exempt from owned-process monitoring because its trainer process is externally owned.

## Change structure and scope

The accepted-result-source contract is Client API-specific, while part of the implementation hardens shared job-process lifecycle infrastructure that is used by every CJ and SJ:

1. **Client API-specific source lifecycle.** The Client API protocol and backends track an accepted lazy result source until receiver settlement, detect owned trainer exit/disconnect, notify the exact DownloadService receiver, interrupt the affected reference without entering `DOWNLOAD_RETRY`, distinguish explicit run abort from source failure, and clean active and retired `launch_once=false` trainer groups. Attach sessions close without killing their externally owned trainer.
2. **Shared CJ/SJ lifecycle prerequisites.** Generic command-callback admission and the job-process teardown pipeline reject new callbacks, archive while F3 is still available, stop F3/Cell, drain admitted callbacks, and close audit/security state in that order. Generic child-outcome delivery reports an authoritative process result before logout, and a logically `STOPPED` but still-live child remains reclaimable. These races predate Client API, but the accepted-source fatal path exercises them deterministically; removing this cluster breaks the failure contract above.
3. **Generic server completion hardening.** Workflow submission admission is serialized with server-runner completion so late task/result callbacks cannot enter already-finalized state.

The deliberate behavior change is that loss of an already accepted lazy source is run-fatal: it reaches `FINISHED:EXECUTION_EXCEPTION` even when controller response thresholds might otherwise tolerate a missing client. A user-requested abort remains authoritative and reaches `FINISHED:ABORTED`.

Crash-safe tensor scratch reclamation after abrupt SJ/CJ death is not implemented here. This PR guarantees owned trainer-process-group cleanup after abrupt CJ death, but not reclamation of a tensor offload directory owned only by a killed job process. The removed parent-tempdir discovery approach was local-launcher-only, could not cross Docker/Kubernetes mount namespaces, and tied reclamation to one parent PID. The P0 follow-up will use launcher-provisioned, parent-owned job scratch whose cleanup capability survives process/container boundaries and parent restart.

### Late accepted Client API results

For `launch_once=false`, a trainer that publishes a late result can outlive its owning Client Job process when the server has already accepted enough responses. The result-source reaper could wait longer than the outer CJ teardown budget, allowing the CJ to exit first and leave the trainer process group orphaned.

If an owned trainer crashes or is OOM-killed after `RESULT_READY` is accepted, its lazy source can disappear while FedAvg or Swarm is still downloading tensors. Previously the receiver kept retrying the permanently unavailable source for the normal streaming timeout, leaving the workflow RUNNING. This PR propagates the exact accepted-source loss through DownloadService and the workflow's existing error path, producing a bounded `FINISHED:EXECUTION_EXCEPTION` result.

### Graceful Swarm abort during tensor offload

A graceful abort can arrive while a remote multi-GB contribution is being materialized under the aggregation client's tensor disk-offload root. Swarm performs that terminal materialization through an explicit FOBS round trip outside `Cell.decode_payload()`; its fresh decode context did not carry the run abort signal. Cleanup waited only for controller-owned learning/aggregation threads, while the active download could be writing from an AUX/streaming handler thread.

An external trainer whose lazy result had already been accepted can also outlive the nested workflow task that originally launched it. In that state `ABORT_TASK` may no longer reach the backend, and stopping only the receiver leaves the producer waiting for its ordinary accepted-source drain. The receiver therefore needs a transaction-level way to report that it abandoned an acquired download.

## Lifecycle contract

Normal END_RUN gives an accepted live result source one short acknowledged SHUTDOWN interval so truthful publication can settle, then force-cleans every still-owned process group.

Graceful abort propagates to the active workflow, lazy tensor download, disk consumer, and external trainer. Once a receiver has learned that its producer supports cancellation, abandoning the download sends terminal receiver failure back to the producer. The producer applies that failure to every ref for the same receiver in the transaction, allowing the accepted source to settle without the normal drain timeout. Cancellation is identity- and acquisition-checked; mixed-version peers retain the existing bounded timeout fallback.

An accepted-source loss is run-fatal: it interrupts the matching download and uses the controller's failure path even when a controller could otherwise tolerate fewer responses. A user-requested run abort remains authoritative over a racing process failure.

Root deletion occurs only after active consumers are cancelled and in-flight chunk writes leave their serialized critical section. Source-loss panic follows normal workflow teardown, so live server/client job processes remove their own roots before exit.

## Validation

- Current PR head: signed `36e5ff529283d0c3c827c010f443256e8cf4132b` (tree `a1d21aa33abf339c1d3114b39381d63af3bf984e`).
- GitHub CI is green on the current head, including Linux unit matrices for Python 3.10 through 3.14 on Ubuntu 22.04 and 24.04, coverage, style, CodeQL, wheel build, documentation links, skill validation, and maintained Hugging Face compatibility checks.
- Current split: 46 focused lifecycle/offload regressions passed.
- Current affected-area suite on the merged head: 1,140 passed, 1 skipped.
- Full repository Black, isort, flake8, and agent-skill checks passed.
- Unit coverage includes duplicate/idempotent cancellation, multi-receiver and sibling-ref accounting, cancellation racing an in-flight serve, first-reply and pipelined aborts, source connection loss, consumer I/O failure, failed cancel delivery, normal completion without cancellation, SIGKILL of an accepted-result trainer, authoritative RC-file preservation, bounded STOPPED-worker cleanup grace, and source-loss panic reaching END_RUN/root removal.
- Current-head GPU system qualification is still in progress. The unchanged maintained `client_api` manifest has one observed failure in `lightning-scaffold-client-api`: both DDP rank-1 processes receive `api=None` during `flare.init()`. The same failure reproduced on predecessor PR head `5db3b3b2`; an unchanged comparison on merged-main parent `be2c91b733440aeb0eebe3191b10932e2d18a3e9` is queued to classify whether it is pre-existing or introduced by this PR. This item is not claimed as passed.
- Exact signed predecessor-head lifecycle E2E: trainer SIGKILL after accepted lazy result and active disk download reached `FINISHED:EXECUTION_EXCEPTION` in 9.31 seconds, preserved/consumed RC101, emitted the authenticated failure report, interrupted the exact download without `DOWNLOAD_RETRY`, fired all END_RUN events, and removed job processes and roots.
- Prior exact signed-head E2E: graceful active-transfer abort reached `FINISHED:ABORTED` with all required END_RUN events and no process, root, descriptor, socket, or GPU residue through delayed audits.
- Exact signed predecessor-head normal-control E2E: full six-round Swarm completion and a separate subsequent canary both reached `FINISHED:COMPLETED` and cleaned resources.
- Abrupt SJ/CJ scratch-directory reclamation and cross-launcher scratch ownership are explicitly out of scope for this PR and must be qualified in the launcher-owned scratch follow-up.

Fixes #5099.

### CODE REVIEW

File: nvflare/utils/tensor_disk_offload.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            ):
+                continue
+            # Close the marker before deletion and revalidate the root identity to
+            # avoid following a replacement at the final mutation boundary.
+            final_stat = os.stat(entry.path, follow_symlinks=False)
+            if not stat.S_ISDIR(final_stat.st_mode):
+                continue
+            if (final_stat.st_dev, final_stat.st_ino) != (root_stat.st_dev, root_stat.st_ino):
+                continue
+            shutil.rmtree(entry.path)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/utils/tensor_disk_offload.py]: Agreed. Rather than extending this pathname-based parent sweep, commit `5db3b3b27ea8fc977d93cc9cc5c19a0854442aa5` removes `nvflare/utils/tensor_disk_offload.py` and both parent-side discovery/reclaimer call sites from #5097. Workflow-local cleanup during normal CJ/SJ teardown remains. Crash-safe cleanup after abrupt job-process death is now a separate P0 design using launcher-provisioned, parent-owned scratch, so this validated-pathname/`shutil.rmtree` race is no longer present in this PR.

File: nvflare/fuel/f3/streaming/download_service.py [hunk 2]
```diff
@@ -1118,8 +1155,141 @@ def _initialize(cls, cell: Cell):
...
+        cell: Cell,
+        targets,
+        source_fqcn: str,
+        ref_ids,
+        reason: str,
+        secure: bool = False,
+    ):
+        """Notify exact downstream receivers that an owned download source died."""
+        if FQCN.validate(source_fqcn):
+            raise ValueError(f"invalid source_fqcn {source_fqcn!r}")
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/fuel/f3/streaming/download_service.py]: Fixed. `notify_source_failure` now stores the result of `FQCN.validate(source_fqcn)` and includes it in the `ValueError`: `invalid source_fqcn ...: <validation error>`. `test_notify_source_failure_reports_fqcn_validation_error` verifies the underlying `invalid char` diagnostic is preserved and that no request is sent.

File: nvflare/utils/tensor_disk_offload.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        root_quarantined = False
+    except BaseException as cleanup_error:
+        if root_quarantined:
+            try:
+                # A transient recursive-delete failure must not strand a hidden
+                # quarantine forever. Put the same inode back at its discoverable
+                # job-scoped path so a later parent cleanup can retry. If another
+                # object has appeared there, preserve the quarantine rather than
+                # overwrite an unrelated path.
+                os.rename(root_name, root_dir, src_dir_fd=cleanup_fd)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/utils/tensor_disk_offload.py]: Agreed. The no-replace rollback evolved into substantial shared-tempdir ownership machinery, and the broader review showed that the approach was still local-launcher-only and tied to one parent PID. Commit `5db3b3b27ea8fc977d93cc9cc5c19a0854442aa5` removes the entire marker/quarantine/rollback mechanism from #5097, so it can no longer replace a competing destination. Abrupt-worker scratch reclamation moves to a separate launcher-owned scratch P0 follow-up.

File: nvflare/utils/tensor_disk_offload.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    root_quarantined = False
+    expected_identity = (expected_stat.st_dev, expected_stat.st_ino)
+    try:
+        cleanup_fd = _open_directory(cleanup_dir)
+        search_fd = _open_directory(search_root)
+        opened_cleanup_stat = os.fstat(cleanup_fd)
+        if (opened_cleanup_stat.st_dev, opened_cleanup_stat.st_ino) != (cleanup_stat.st_dev, cleanup_stat.st_ino):
+            raise RuntimeError("tensor disk-offload quarantine identity changed before use")
+        root_name = os.path.basename(root_dir)
+        os.rename(root_dir, root_name, dst_dir_fd=cleanup_fd)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/utils/tensor_disk_offload.py]: Agreed. Commit `5db3b3b27ea8fc977d93cc9cc5c19a0854442aa5` removes the pathname-discovery and quarantine mechanism instead of trying to close another TOCTOU window around it. #5097 now retains only workflow-local root cleanup reached through normal teardown. The separate crash-safe scratch design must provision ownership at the launcher boundary and retain a cleanup capability across process/container boundaries, rather than rediscovering mutable `/tmp` paths.

File: nvflare/utils/tensor_disk_offload.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    finally:
+        if cleanup_fd >= 0:
+            os.close(cleanup_fd)
+        if search_fd >= 0:
+            os.close(search_fd)
+        if not root_quarantined:
+            # This removes only the now-empty quarantine directory. It is never
+            # used for recursive deletion, so a replacement cannot expose an
+            # unrelated tree to cleanup.
+            os.rmdir(cleanup_dir)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/utils/tensor_disk_offload.py]: Agreed. The quarantine final-removal path, including this descriptor-close/pathname-reopen race and exception masking, was removed in commit `5db3b3b27ea8fc977d93cc9cc5c19a0854442aa5` along with `nvflare/utils/tensor_disk_offload.py`. The PR description and design doc now state the reduced scope; abrupt CJ/SJ scratch reclamation is a separate launcher-owned follow-up.

**Comment by pcnudde** [comment]
When PID-based owner monitoring disables itself at session start (containerized trainer, CJ PID not visible), the remaining backstop is the CJ-heartbeat timeout — but that path only cooperatively aborts the session (`_abort_if_cj_timed_out` sets flags and triggers signals); it never calls `_terminate_orphaned_process_group`, which is reachable only from the disabled watchdog loop.

So after a hard-killed CJ, the trainer main process exits only if the training script's next flare call raises and propagates. A script blocked in a long GPU collective (or one catching exceptions) never exits, and DDP/dataloader child processes are never signaled in any path — orphaned processes keep holding GPUs in exactly the docker/k8s deployments where PID visibility is lost.

Suggested fix: in the heartbeat-timeout path, after the cooperative abort, escalate to the existing `_terminate_orphaned_process_group()` (gated on non-attach so externally owned trainers are untouched, optionally after a short grace so a well-behaved script can exit cleanly first).

**Review by pcnudde — APPROVED** [review]
